File: apps/gateway/src/gateway/routes/playbooks.py
from __future__ import annotations

import uuid
from pathlib import Path

import yaml
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from shared.schemas import Playbook

router = APIRouter()

# In-memory store for user-saved playbooks
_user_playbooks: dict[str, Playbook] = {}

def load_seed_playbooks() -> list[Playbook]:
    seed_file = Path(__file__).parent.parent.parent.parent.parent / "data" / "seed-playbooks.yaml"
    if not seed_file.exists():
        return []
    with open(seed_file, "r") as f:
        data = yaml.safe_load(f)
    return [Playbook(**item) for item in data] if data else []

@router.get("", response_model=list[Playbook])
async def list_playbooks() -> list[Playbook]:
    seeds = load_seed_playbooks()
    return seeds + list(_user_playbooks.values())

class SavePlaybookRequest(BaseModel):
    session_id: str
    title: str
    description: str

@router.post("", response_model=Playbook)
async def save_playbook(request: SavePlaybookRequest) -> Playbook:
    # In a real implementation, we would extract the prompt from the session trace.
    # For now, we'll just use a placeholder prompt.
    playbook = Playbook(
        id=str(uuid.uuid4()),
        title=request.title,
        description=request.description,
        intent="Custom",
        prompt_template="Custom prompt from session " + request.session_id,
        created_by="you",
        created_at="2026-05-19T00:00:00Z",
    )
    _user_playbooks[playbook.id] = playbook
    return playbook

@router.get("/{playbook_id}", response_model=Playbook)
async def get_playbook(playbook_id: str) -> Playbook:
    seeds = load_seed_playbooks()
    for seed in seeds:
        if seed.id == playbook_id:
            return seed
    if playbook_id in _user_playbooks:
        return _user_playbooks[playbook_id]
    raise HTTPException(status_code=404, detail="Playbook not found")

File: apps/web/lib/playbooks.ts
export interface Playbook {
  id: string;
  title: string;
  description: string;
  intent: string;
  prompt_template: string;
  created_by: string;
  created_at: string;
}

export interface SavePlaybookRequest {
  session_id: string;
  title: string;
  description: string;
}

const BASE = '/api';

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    headers: { 'Content-Type': 'application/json' },
    ...init,
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail?.message ?? `HTTP ${res.status}`);
  }
  return res.json();
}

export async function listPlaybooks(): Promise<Playbook[]> {
  return request('/playbooks');
}

export async function getPlaybook(id: string): Promise<Playbook> {
  return request(`/playbooks/${id}`);
}

export async function savePlaybook(data: SavePlaybookRequest): Promise<Playbook> {
  return request('/playbooks', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}
